### PR TITLE
Transfer `blob_store` into this repo

### PR DESCRIPTION
**Pull request by coszio**
This huge PR transfers the entirety of [`blob_store`](https://github.com/qdrant/mmap-payload-storage) into this main repo.

The copied utilities have been replaced, and the workspace dependency versions are preferred.

### Tests times
In the repo, there are tests which take some considerable time to run. A quick measurement:
Total time before: 136.308s
Total time now: 159.941s

While it is not too dramatic, it is does increase it considerably. Some of the tests can be modified to take less time at the expense of not being so thorough. Open for discussion.

### CODE REVIEW

File: lib/blob_store/data/h&m-articles.csv
Comment (timvisee) [inline lib/blob_store/data/h&m-articles.csv:1]: I also vote to remove this from the source tree, as discussed in Slack.

File: lib/blob_store/src/blob_store.rs [hunk 1]
```diff
@@ -0,0 +1,983 @@
...
+                    let payload = random_payload(rng, size_factor);
+                    Operation::Update(point_offset, payload)
+                }
+                _ => unreachable!(),
+            }
+        }
+    }
+
+    #[rstest]
+    fn test_behave_like_hashmap(
```
Comment (agourlay) [inline lib/blob_store/src/blob_store.rs:681]: I vote for generating less operations as the tests are really heavy

```console
        SLOW [>120.000s] blob_store blob_store::tests::test_behave_like_hashmap::page_size_1_1_048_576
        SLOW [>120.000s] blob_store blob_store::tests::test_behave_like_hashmap::page_size_2_2_097_152
        SLOW [>120.000s] blob_store blob_store::tests::test_behave_like_hashmap::page_size_3_DEFAULT_PAGE_SIZE_BYTES
        PASS [ 123.406s] blob_store blob_store::tests::test_behave_like_hashmap::page_size_3_DEFAULT_PAGE_SIZE_BYTES
```

File: lib/blob_store/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+edition = "2021"
+publish = false
+
+[lints]
+workspace = true
+
+[features]
+bench_rocksdb = ["dep:rocksdb"]
+
+[dependencies]
```
Comment (agourlay) [inline lib/blob_store/Cargo.toml:15]: cargo machete says there are unused deps

```
        atomicwrites
        bincode
        log
        nix
        thiserror
```

Comment (coszio) [reply lib/blob_store/Cargo.toml:15]: Thanks for running it, I forgot to remove these when removing the copied utilities

File: lib/blob_store/src/tracker.rs [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+    /// Path to the file
+    path: PathBuf,
+    /// Header of the file
+    header: TrackerHeader,
+    /// Mmap of the file
+    mmap: MmapMut,
+    /// Updates that haven't been flushed
+    ///
+    /// When flushing, these updates get written into the mmap and flushed at once.
+    pending_updates: HashMap<PointOffset, PointerUpdate>,
```
[the remark this answers is not in the document]
Comment (coszio) [reply lib/blob_store/src/tracker.rs:78]: We should change this in a separate PR

**Review by agourlay — APPROVED** [review]

**Pull request merged**